cpp/src/arrow/flight/client.cc

```diff
@@ -620,6 +621,17 @@ Status FlightClient::GetFlightInfo(const FlightCallOptions& options,
   return GetFlightInfo(options, descriptor).Value(info);
 }
 
+void FlightClient::GetFlightInfo(const FlightCallOptions& options,
+                                 const FlightDescriptor& descriptor,
+                                 std::shared_ptr<AsyncListener<FlightInfo>> listener) {
+  if (auto status = CheckOpen(); !status.ok()) {
+    listener->OnFinish(
+        TransportStatus{TransportStatusCode::kInternal, status.ToString()});
```

Comment:
It may be better that we don't export `TransportStatus` to users because we have many useful macros for `arrow::Status` such as `ARROW_RETURN_NOT_OK()`.
Can we use `arrow::Status` directly?

It seems that `TransportStatus` may have `arrow::Status` via `ArrowStatusDetail`.
Can we put `TransportStatus` into `arrow::Status::detail()` by creating `TransportStatusDetail` or something?

Reply:
It's intentional; I've come to think that it was a mistake to try to map the RPC errors onto the Arrow statuses, because they don't match, and because they carry different kinds of metadata (e.g. Arrow Status only allows one detail with a type while gRPC status is an array of binary blobs).

ArrowStatusDetail is a way to propagate internal errors from the client (not so useful here, but for example if a server sends an invalid Arrow record batch).

And because the status is given to you in a callback, the macros aren't useful anyways. (You can still convert the TransportStatus to an Arrow Status if you want to use arrow::Future, for instance.)

Reply:
Thinking about it some more: maybe the TransportStatusDetail would be preferable? If it's missing, then we know it's an internal error, otherwise we know it's an RPC error. I think you are right that consistency with the rest of Arrow is important, but I just don't like how synchronous Flight RPC makes it hard to recover the original RPC error code.

Reply:
And TransportStatusDetail could be 

```
class TransportStatusDetail : public StatusDetail {
  TransportStatusCode code;
  std::vector<ErrorDetail> error_details;
};
// error detail is pair<string, variant<string, bytes>> or something along those lines (or basically, HTTP headers)
```

I wish we could use `FlightStatusDetail` but that name is taken

Reply:
I see. I agree with you.

Reply:
At this point the original implementation is too much of a mess or we would have to make a potentially breaking change (we'd have to replace or deprecate FlightStatusDetail and/or the mapping between Arrow and Flight status codes) that has to be coordinated across client/server APIs (since the existing FlightStatusCode is also used to return certain status codes)